src/dictionary/linear_hash/linear_hash.c

```diff
@@ -0,0 +1,1364 @@
+						return status.error;
+					}
+				}
+			}
+
+			record_loc += record_total_size;
+		}
+	}
+
+	return linear_hash_increment_next_split(linear_hash);
+}
+
+int
+linear_hash_above_threshold(
+	linear_hash_table_t *linear_hash
+) {
+	double	numerator	= (double) (100 * (linear_hash->num_records));
+	double	denominator = (double) (linear_hash->num_buckets * linear_hash->records_per_bucket);
+
+	double load			= numerator / denominator;
+
+	int above_threshold = (load > linear_hash->split_threshold);
+
+	return above_threshold;
+}
+
+/* TODO it may not be necesarry to copy the key to a location */
+/* linear hash operations */
+ion_status_t
+linear_hash_insert(
+	ion_key_t			key,
+	ion_value_t			value,
+	int					hash_bucket_idx,
+	linear_hash_table_t *linear_hash
+) {
+	ion_status_t status = ION_STATUS_INITIALIZE;
+
+	if (hash_bucket_idx < linear_hash->next_split) {
+		hash_bucket_idx = hash_to_bucket(key, linear_hash);
+	}
+
+	/* create a linear_hash_record with the desired key, value, and status of full*/
+	ion_byte_t	*record_key		= alloca(linear_hash->super.record.key_size);
+	ion_byte_t	*record_value	= alloca(linear_hash->super.record.value_size);
+	ion_byte_t	record_status	= 1;
+
+	memcpy(record_key, key, linear_hash->super.record.key_size);
+	memcpy(record_value, value, linear_hash->super.record.value_size);
+
+	ion_fpos_t record_total_size		= linear_hash->super.record.key_size + linear_hash->super.record.value_size + sizeof(ion_byte_t);
+
+	/* get the appropriate bucket for insertion */
+	ion_fpos_t				bucket_loc	= bucket_idx_to_ion_fpos_t(hash_bucket_idx, linear_hash);
+	linear_hash_bucket_t	bucket;
+
+	status.error = linear_hash_get_bucket(bucket_loc, &bucket, linear_hash);
+
+	if (status.error != err_ok) {
+		return status;
+	}
+
+	/* location of the records in the bucket to be stored in */
+	ion_fpos_t	bucket_records_loc = get_bucket_records_location(bucket_loc);
+	ion_fpos_t	record_loc;
+
+	/* Case the bucket is empty */
+	if (bucket.anchor_record == -1) {
+		record_loc				= bucket_records_loc;
+		bucket.anchor_record	= record_loc;
+	}
+	else {
+		/* Case that the bucket is full but there is not yet an overflow bucket */
+		if (linear_hash_bucket_is_full(bucket, linear_hash)) {
+			/* Get location of overflow bucket and update the tail record for the linked list of buckets storing
+			 * items that hash to this bucket and update the tail bucket with the overflow's location */
+			ion_fpos_t *overflow_location = alloca(sizeof(ion_fpos_t));
+
+			status.error = create_overflow_bucket(bucket.idx, overflow_location, linear_hash);
+
+			if (err_ok != status.error) {
+				return status;
+			}
+
+			/* Set the location of the anchor record on the new overflow bucket and update the record_loc for storing
+			 * the new record to be this location */
+			ion_fpos_t overflow_anchor_record_loc = get_bucket_records_location(*overflow_location);
+
+			status.error = linear_hash_get_bucket(*overflow_location, &bucket, linear_hash);
```

Comment:
no, good catch! the update below was not necessary either as it was handled below.